### PR TITLE
Update emojis and use sprite sheet instead of individual images

### PR DESCRIPTION
**Pull request by rubengees**
This is a rewrite of the generator for updating the emojis and switching to the sprite sheet approach discussed in https://github.com/vanniktech/Emoji/issues/88.

We now use [this](https://github.com/iamcal/emoji-data) data source instead of the EmojiOne metadata file, which does not seem to get updated anytime soon.

I have marked this as WIP, as I'm not sure if this is the right way to go. I also have not regenerated yet to make this PR reviewable.

### Wins

- More emojis! :1st_place_medal:
- Size reduction. Not too great but notable. `Ios` went from 4.5MB to 3.7MB.

### Important code changes

A `<Name>Emoji` class is now generated for each emoji type (e.g. `IosEmoji`). This is a subclass of `Emoji` to make the sprite sheet approach work. It works by loading the big sheet into a static field and then cuts the individual emojis from that when requested.

The generator does not download any files anymore but depends on the aforementioned library through `npm` now.

### Points of discussion

- The ordering is different. It makes sense but is different from WhatsApp (which we always tried to copy in previous versions). It is now based on the official [unicode recommendations](https://www.unicode.org/emoji/charts/emoji-ordering.html). Slack uses this ordering now also for example.

- Once the big sheet is loaded, there is no way to reclaim that memory. Should we introduce a way of uninstalling a `Provider`? We then would also need a mechanism to destroy `Emoji` instances. Currently I'm thinking of making the `destroy` method of the `EmojiManager` public and adding a `destroy` method to `Emoji` which subclasses can override.

### CODE REVIEW

**Comment by rubengees** [comment]
- The images are already heavily optimized ([pngquant](https://pngquant.org/) and [zopfli](https://github.com/google/zopfli)). That is the point of having one large image I think - You can optimize that pretty well.

- Yeah thought about that as well. We now have no `Provider` which uses the old logic from the `Emoji` class. I could imagine though that somebody is using that for a custom `Provider`. But I would be fine with converting to an interface.

- Sorry, totally forgot about that, here they are:
<details><summary>Click to expand</summary><p>

bildschirmfoto von 2018-02-04 13-29-08
bildschirmfoto von 2018-02-04 13-30-44 
bildschirmfoto von 2018-02-04 13-30-54
bildschirmfoto von 2018-02-04 13-31-04
bildschirmfoto von 2018-02-04 13-31-10
bildschirmfoto von 2018-02-04 13-31-17
bildschirmfoto von 2018-02-04 13-31-23
bildschirmfoto von 2018-02-04 13-31-30
bildschirmfoto von 2018-02-04 13-31-37
bildschirmfoto von 2018-02-04 13-31-47
bildschirmfoto von 2018-02-04 13-32-01
</p></details><br>

- Yes I actually have! In one app I work on, I constantly fight with OEMs and other problems related to low memory since that app is really image-heavvy. I would rather not have around 4MB loaded at all time but reclaim that memory when the user leaves the `Activity` with the emoji picker. I can imagine there are more people with similar problems...

**Comment by vanniktech** [comment]
> The images are already heavily optimized (pngquant and zopfli). That is the point of having one large image I think - You can optimize that pretty well.

Nice.

> Yeah thought about that as well. We now have no Provider which uses the old logic from the Emoji class. I could imagine though that somebody is using that for a custom Provider. But I would be fine with converting to an interface.

Then leave things as is in this PR to minimize the changes.

> Sorry, totally forgot about that, here they are:

Oh wow things are really sorted differently. Is there any way to get a sorting that's close to Whatsapp ones? From the user perspective it's best if we use the same ordering otherwise you always need to end up looking for emojis which is the worst thing ever. Also we don't have any search functionality yet so this becomes even more important.

> Yes I actually have! In one app I work on, I constantly fight with OEMs and other problems related to low memory since that app is really image-heavvy. I would rather not have around 4MB loaded at all time but reclaim that memory when the user leaves the Activity with the emoji picker. I can imagine there are more people with similar problems...

Then let's do it.

**Comment by rubengees** [comment]
> From the user perspective it's best if we use the same ordering otherwise you always need to end up looking for emojis which is the worst thing ever.

I agree! Sadly I don't know of a way to get the WhatsApp ordering with the data we have. We require a metadata file with categories and ordering (like that in WhatsApp) that currently only that file from EmojiOne has (at least I have not found another one). And that one does not get updated sadly.

Is this PR then out of question? I have looked around and there are at least two apps that have a somewhat similar ordering: Slack and the Telegram X app.

**Comment by vanniktech** [comment]
> Is this PR then out of question?

No not at all. Let's ditch the order then and possibly try to tweak it. The benefits here outweigh that entirely.

**Comment by rubengees** [comment]
I have added a machanism to reclaim memory. Here is a screenshot of it in action (Android Profiler):

bildschirmfoto von 2018-02-05 22-04-23

> possibly try to tweak it

I can't think of a non-manual approach, but will keep looking around.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 1]
```diff
@@ -134,11 +134,16 @@ public static void install(@NonNull final EmojiProvider provider) {
     INSTANCE.emojiRepetitivePattern = Pattern.compile('(' + regex + ")+");
   }
 
-  static void destroy() {
+  public static void destroy() {
+    for (final Emoji emoji : INSTANCE.emojiMap.values()) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: Let's use fori here too

Comment (rubengees) [reply emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: We can't; [values()](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#values()) returns a `Collection` and not a `List`.

**Comment by rubengees** [comment]
Adjusted a few things (the first two commits) and regenerated!

**Comment by vanniktech** [comment]
Thanks. I'll take a look later.

**Pull request merged**